### PR TITLE
Stop/down clusters with new provisioner

### PR DESCRIPTION
**Pull request by suquark**
This PR is part of the new provisioner PR, which handles stopping/terminating of clusters. This PR only includes the implementation for AWS, but it can serve as a starting point for other clouds.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [x] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/backends/cloud_vm_ray_backend.py [hunk 1]
```diff
@@ -3281,6 +3282,22 @@ def teardown_no_lock(self,
...
+            #  "clouds.AWS" here). Currently I just use the code from
+            #  the new provisioner, which I believe benefits further
+            #  development of "SkyPilot as a service". Let's see how
+            #  this goes.
+            region = config['provider']['region']
+            provision_lib.teardown_cluster(repr(cloud), region, cluster_name,
+                                           terminate)
+            if post_teardown_cleanup:
+                self.post_teardown_cleanup(handle, terminate, purge)
+            return
```
Comment (Michaelvll) [inline sky/backends/cloud_vm_ray_backend.py:3299]: This is great!
More I am thinking about this, I am wondering if we can have all those operations that reach the cloud providers using their APIs to be located within the `provision` module (may need to rename the `provision` to another name, if we decide to go that way). That said, I am thinking to move the `cloud.query_status` to `provision` module as well, so that to leave the `Cloud` as a frontend class for operating/querying the catalogs, and `provision` module to be the backend for each cloud.

File: sky/provision/aws/utils.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+                    region: str,
+                    max_attempts: int = BOTO_MAX_RETRIES,
+                    **credentials) -> Any:
+    """Create an AWS resource."""
+    # overhead: 6.69 ms ± 41.5 µs
+    return aws.resource(
+        resource,
+        region=region,
+        config=config.Config(retries={'max_attempts': max_attempts}),
+        **credentials)
```
Comment (Michaelvll) [inline sky/provision/aws/utils.py:20]: This function is a bit shallow. Should we just move this `max_attempts` argument into the `adaptors.aws.resource` function?

File: sky/provision/aws/instance.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        # Waiting the cluster results in errors (cannot wait for 'terminated').
+        # So here we exclude terminated/terminating instances.
+        filters.append({
+            'Name': 'instance-state-name',
+            'Values': ['pending', 'running'],
+        })
+    elif state == common.InstanceState.STOPPED:
+        filters.append({
+            'Name': 'instance-state-name',
+            'Values': ['stopping', 'stopped'],
```
Comment (Michaelvll) [inline sky/provision/aws/instance.py:114]: Question: why we only wait for instance with `stopping` and `stopped` states for `state == common.InstanceState.STOPPED`, but waiting for all the instances when `state == common.InstanceState.TERMINATED`? Should we wait for all the instance except for the `terminating` and `terminated` ones to be stopped? Similar for the `RUNNING`, should we wait for the `stopped` `stopping` instances as well?

File: sky/backends/cloud_vm_ray_backend.py [hunk 4]
```diff
@@ -3281,6 +3282,22 @@ def teardown_no_lock(self,
         config = common_utils.read_yaml(handle.cluster_yaml)
         cluster_name = handle.cluster_name
         use_tpu_vm = config['provider'].get('_has_tpus', False)
+
+        # Use the new provisioner for AWS.
+        if isinstance(cloud, clouds.AWS):
```
Comment (Michaelvll) [inline sky/backends/cloud_vm_ray_backend.py:3287]: We also call a `ray down` in the following place, do we want to replace that as well? Or do you think that can be done in the future PR as that needs to handle the separation of `head` and `worker` nodes?

https://github.com/skypilot-org/skypilot/blob/db5d15b4269db5dbeeebdc87cca357a2fc620c00/sky/skylet/events.py#L166

File: sky/provision/aws/instance.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    ec2 = utils.create_resource('ec2', region=region)
+    filters = [
+        {
+            'Name': 'instance-state-name',
+            'Values': ['pending', 'running'],
+        },
+        {
+            'Name': f'tag:{TAG_RAY_CLUSTER_NAME}',
+            'Values': [cluster_name],
+        },
```
Comment (Michaelvll) [inline sky/provision/aws/instance.py:56]: For this `ray` specific filtering, I am thinking maybe we can move it upper to the `backend.provision_lib` and have this `stop_instances` having the following signature:
```
def stop_instances(region: str,
                   cluster_name: str,
                   tag_filters: Dict[str, str],
                   excluded_instances: Optional[List[str]] = None) -> None:
```
The include/exclude may also be done with the `tag_filters`, just as what we did for the `cloud.query_status`?

Reason for this change: I am wondering if we can make the `provision` module decoupled with the ray-specific implementation. All the ray backend related things can go under `backends/`. Wdyt?

**Review by Michaelvll — COMMENTED** [review]
Thanks for separating this PR out from the new provisioner PR @suquark! I really like how tidy these codes look. Left several questions and thought, mainly about how we decouple things. We can jump on a call when more discussion is needed.

File: sky/provision/aws/instance.py (hunk 3 above)
Comment (Michaelvll) [reply sky/provision/aws/instance.py:114]: Got it! That makes sense. Thanks for the explanation

File: sky/backends/cloud_vm_ray_backend.py [hunk 6]
```diff
@@ -3281,6 +3282,25 @@ def teardown_no_lock(self,
...
+            # TODO(suquark): An alternative approach is to move cloud
+            #  termination logic into cloud classes (for example,
+            #  "clouds.AWS" here). Currently I just use the code from
+            #  the new provisioner, which I believe benefits further
+            #  development of "SkyPilot as a service". Let's see how
+            #  this goes.
```
Comment (Michaelvll) [inline sky/backends/cloud_vm_ray_backend.py:3293]: This might be staled?

File: sky/backends/cloud_vm_ray_backend.py [hunk 7]
```diff
@@ -3281,6 +3282,25 @@ def teardown_no_lock(self,
...
+            region = config['provider']['region']
+            if not self.run_on_head(handle, 'ray stop --force'):
```
Comment (Michaelvll) [inline sky/backends/cloud_vm_ray_backend.py:3295]:
```suggestion
            # Stop the ray autoscaler first to avoid the head node trying to
            # re-launch the worker nodes, during the termination of the
            # cluster.
            if not self.run_on_head(handle, 'ray stop --force'):
```

File: sky/backends/cloud_vm_ray_backend.py [hunk 8]
```diff
@@ -3281,6 +3282,25 @@ def teardown_no_lock(self,
...
+                logger.warning(
+                    'failed to take down Ray autoscaler on the head node')
```
Comment (Michaelvll) [inline sky/backends/cloud_vm_ray_backend.py:3297]:
```suggestion
                    'Failed to take down Ray autoscaler on the head node. '
                    'It might be because the cluster's head node has already '
                    'been terminated. It is fine to skip this.')
```

File: sky/backends/provision_lib.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+def teardown_cluster(cloud_name: str, region: str, cluster_name: str,
+                     terminate: bool) -> None:
+    """Deleting or stopping a cluster."""
+    if terminate:
+        provision.terminate_instances(cloud_name, region, cluster_name)
+        # TODO(suquark): enable this line in later PRs about the
+        #  new provisioner
+        # metadata_utils.remove_cluster_metadata(cluster_name)
+    else:
+        provision.stop_instances(cloud_name, region, cluster_name)
```
Comment (Michaelvll) [inline sky/backends/provision_lib.py:17]: Question: do we really need this module, as it only serves as a router based on `terminate` argument, and it does not seem to be a complicated enough function to have this router. Should we directly let the `backend` call the `provision.terminate_instances` and `provision.stop_instances`?

File: sky/backends/provision_lib.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+    # TODO(suquark): Currently, the implementation of GCP and Azure will
+    #  wait util the cluster is fully terminated, while other clouds just
+    #  trigger the termination process (via http call) and then return.
+    #  It's not clear that which behavior should be expected. We will not
+    #  wait for the termination for now, since this is the default behavior
+    #  of most cloud implementations.
```
Comment (Michaelvll) [inline sky/backends/provision_lib.py:24]: Should we move this comment to `aws/instances.py` 's module docstr?

File: sky/provision/aws/instance.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+from typing import Dict, List, Any, Optional
+
+from botocore import config
+from sky.adaptors import aws
+
+# ======================== About AWS subnet/VPC ========================
+# https://stackoverflow.com/questions/37407492/are-there-differences-in-networking-performance-if-ec2-instances-are-in-differen
+# https://docs.aws.amazon.com/vpc/latest/userguide/how-it-works.html
+# https://docs.aws.amazon.com/vpc/latest/userguide/configure-subnets.html
+
```
Comment (Michaelvll) [inline sky/provision/aws/instance.py:11]: Seems not related to the functions in this module?

File: sky/provision/aws/instance.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+# ======================== Instance state and lifecycle ========================
+# https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-lifecycle.html
+
+# ======================== About AWS availability zone ========================
+# Data transfer within the same region but different availability zone
+#  costs $0.01/GB:
+# https://aws.amazon.com/ec2/pricing/on-demand/#Data_Transfer_within_the_same_AWS_Region
```
Comment (Michaelvll) [inline sky/provision/aws/instance.py:18]: Seems not related to this module?

File: sky/provision/aws/instance.py [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+        {
+            'Name': 'instance-state-name',
+            'Values': ['pending', 'running'],
+        },
+        {
+            'Name': f'tag:{TAG_RAY_CLUSTER_NAME}',
+            'Values': [cluster_name],
+        },
+    ]
+    # TODO(suquark): wait pending clusters?
```
Comment (Michaelvll) [inline sky/provision/aws/instance.py:62]: Can we copy the comment in `provision_lib` to here to make easier to refer to?

**Review by Michaelvll — COMMENTED** [review]
Thanks for the updates @suquark! The code mostly looks good to me. While I met the following error during the test, i.e. the `self.run_on_head` does not work well when the cluster does not exist.

```
  File "/home/gcpuser/skypilot/sky/backends/cloud_vm_ray_backend.py", line 3295, in teardown_no_lock
    if not self.run_on_head(handle, 'ray stop --force'):
  File "/home/gcpuser/skypilot/sky/utils/common_utils.py", line 256, in _record
    return f(*args, **kwargs)
  File "/home/gcpuser/skypilot/sky/backends/cloud_vm_ray_backend.py", line 3632, in run_on_head
    head_ip = backend_utils.get_head_ip(handle, _FETCH_IP_MAX_ATTEMPTS)
  File "/home/gcpuser/skypilot/sky/utils/common_utils.py", line 256, in _record
    return f(*args, **kwargs)
  File "/home/gcpuser/skypilot/sky/backends/backend_utils.py", line 1575, in get_head_ip
    head_ip = _query_head_ip_with_retries(handle.cluster_yaml, max_attempts)
  File "/home/gcpuser/skypilot/sky/backends/backend_utils.py", line 1375, in _query_head_ip_with_retries
    raise RuntimeError('Failed to get head ip') from e
RuntimeError: Failed to get head ip
```

**Comment by suquark** [comment]
Thanks for your review! could you also tell me the failed tests? I can test if by myself

File: sky/backends/cloud_vm_ray_backend.py [hunk 14]
```diff
@@ -3310,6 +3311,35 @@ def teardown_no_lock(self,
...
+            # re-launch the worker nodes, during the termination of the
+            # cluster.
+            try:
+                # We do not check the return code, since Ray returns
+                # non-zero return code when calling Ray stop,
+                # even when the command was executed successfully.
+                self.run_on_head(handle, 'ray stop --force')
+            except RuntimeError:
+                # This error is expected if the previous cluster status is
+                # INIT/STOPPED.
```
Comment (Michaelvll) [inline sky/backends/cloud_vm_ray_backend.py:3327]:
```suggestion
                # This error is expected if the previous cluster IP is failed to be
                # found, i.e., the cluster is already stopped/terminated.
```

**Review by Michaelvll — APPROVED** [review]
Thanks for the fix @suquark! Just passed the test smoothly `pytest tests/test_smoke.py`. LGTM.

**Comment by concretevitamin** [comment]
Woohoo!! Very excited to see this @suquark!
